src/charm_state.py

```diff
@@ -5,32 +5,63 @@
 
 import dataclasses
 import logging
+import re
+from typing import Optional
 
 from ops import CharmBase
 
 logger = logging.getLogger(__name__)
 
 COS_AGENT_INTEGRATION_NAME = "cos-agent"
+HOSTNAME_PORT_PATTERN = re.compile(r"^[\w\-.]+:\d+$")
+
+
+class CharmConfigInvalidError(Exception):
+    """Raised when charm config is invalid.
+
+    Attributes:
+        msg: Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CharmConfigInvalidError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
 
 
 @dataclasses.dataclass(frozen=True)
 class State:
     """The charm state.
 
     Attributes:
-        proxy_config: Proxy configuration.
-        _charm: The charm instance.
+        is_metrics_logging_available: Whether the charm is able to issue metrics.
+        aproxy_proxy: The socket address of the proxy to configure aproxy with.
     """
 
     is_metrics_logging_available: bool
+    aproxy_proxy: Optional[str]
 
     @classmethod
     def from_charm(cls, charm: CharmBase) -> "State":
         """Initialize the state from charm.
 
+        Args:
+            charm: The charm instance.
+
         Returns:
             Current state of the charm.
         """
+        if aproxy_proxy := charm.config.get("aproxy-proxy"):
+            if not HOSTNAME_PORT_PATTERN.match(aproxy_proxy):
```

Comment:
Why not using pydantic to validate with a regex?

Reply:
I thought it was a bit simpler than introducing a new class inheriting from `BaseModel` or a pydantic field validator. But we'll probably introduce a ProxySetting class in `charm_state` to when we move the juju proxy setting to that place.

Reply:
Up to you, then